File: cogs/utility.py
"""
Utility cog - Help and utility commands
Mirrors JS utility.js with detailed help categories
"""

import discord
from discord import app_commands
from discord.ext import commands

import wavelink
from typing import cast
import logging
import os
import platform

logger = logging.getLogger('MusicBot.Utility')


def format_duration(ms: int) -> str:
    """Format milliseconds to mm:ss or hh:mm:ss."""
    seconds = ms // 1000
    hours, remainder = divmod(seconds, 3600)
    minutes, seconds = divmod(remainder, 60)
    if hours:
        return f"{hours}:{minutes:02d}:{seconds:02d}"
    return f"{minutes}:{seconds:02d}"


class Utility(commands.Cog):
    """Utility commands."""
    
    def __init__(self, bot: commands.Bot):
        self.bot = bot
    
    @app_commands.command(name="help", description="Show all available commands")
    async def help(self, interaction: discord.Interaction):
        """Display detailed help embed - mirrors JS utility.js help."""
        embed = discord.Embed(
            title="🎵 Music Bot Commands",
            color=discord.Color.from_str('#5865F2'),
            description="A Discord music bot powered by Lavalink"
        )
        
        embed.add_field(
            name="🎶 Playback",
            value="\n".join([
                "`/play` - Play a song or playlist",
                "`/search` - Search with selection",
                "`/playnext` - Add song to play next",
                "`/pause` - Pause playback",
                "`/resume` - Resume playback",
                "`/stop` - Stop and disconnect",
                "`/skip` - Skip current track",
                "`/forceskip` - Force skip (DJ)",
                "`/voteskip` - Vote to skip",
                "`/forceplay` - Play immediately (DJ)",
                "`/previous` - Play previous track",
                "`/seek` - Seek to position",
                "`/volume` - Adjust volume",
                "`/nowplaying` - Show current track",
                "`/loop` - Toggle loop mode",
                "`/replay` - Replay current track",
                "`/lyrics` - Get song lyrics"
            ]),
            inline=False
        )
        
        embed.add_field(
            name="📋 Queue",
            value="\n".join([
                "`/queue` - Show the queue",
                "`/shuffle` - Shuffle the queue",
                "`/clear` - Clear the queue",
                "`/remove` - Remove a track",
                "`/forceremove` - Force remove (DJ)",
                "`/move` - Move a track",
                "`/jump` - Jump to a track",
                "`/skipto` - Skip to a track"
            ]),
            inline=False
        )
        
        embed.add_field(
            name="💾 Playlists",
            value="\n".join([
                "`/playlists` - View saved playlists",
                "`/savelist` - Save queue as playlist",
                "`/loadlist` - Load a playlist",
                "`/appendlist` - Add to playlist",
                "`/deletelist` - Delete a playlist"
            ]),
            inline=False
        )
        
        embed.add_field(
            name="🎛️ Filters",
            value="\n".join([
                "`/filters` - Apply audio filter",
                "`/equalizer` - Equalizer presets",
                "`/speed` - Adjust playback speed",
                "`/pitch` - Adjust audio pitch"
            ]),
            inline=False
        )
        
        embed.add_field(
            name="⚙️ Settings",
            value="\n".join([
                "`/loop` - Toggle loop mode",
                "`/autoplay` - Toggle autoplay",
                "`/247` - Toggle 24/7 mode",
                "`/djrole` - Set DJ role (Admin)",
                "`/serversettings` - View all settings"
            ]),
            inline=False
        )
        
        embed.add_field(
            name="🔧 Admin (Requires Admin/Owner)",
            value="\n".join([
                "`/settc` - Lock bot to text channel",
                "`/setvc` - Lock bot to voice channel",
                "`/queuetype` - Linear or fair queue",
                "`/skipratio` - Vote skip threshold",
                "`/autoplaylist` - Auto-load playlist",
                "`/songinstatus` - Show song in status"
            ]),
            inline=False
        )
        
        embed.add_field(
            name="🔧 Utility",
            value="\n".join([
                "`/help` - Show this message",
                "`/ping` - Check bot latency",
                "`/stats` - Bot statistics"
            ]),
            inline=False
        )
        
        embed.set_footer(text="Powered by Lavalink | DJ commands require DJ role")
        
        await interaction.response.send_message(embed=embed)
    
    @app_commands.command(name="ping", description="Check bot and Lavalink latency")
    async def ping(self, interaction: discord.Interaction):
        """Show bot and Lavalink latency - mirrors JS utility.js ping."""
        await interaction.response.defer()
        
        # Discord latency
        bot_latency = round(self.bot.latency * 1000)
        
        # Lavalink latency
        nodes = wavelink.Pool.nodes
        lavalink_ping = "N/A"
        if nodes:
            node = list(nodes.values())[0]
            if node.heartbeat:
                lavalink_ping = f"{node.heartbeat}ms"
        
        embed = discord.Embed(
            title="🏓 Pong!",
            color=discord.Color.green()
        )
        embed.add_field(name="Bot Latency", value=f"{bot_latency}ms", inline=True)
        embed.add_field(name="WebSocket", value=f"{round(self.bot.latency * 1000)}ms", inline=True)
        embed.add_field(name="Lavalink", value=lavalink_ping, inline=True)
        
        await interaction.followup.send(embed=embed)
    
    @app_commands.command(name="stats", description="Show bot statistics")
    async def stats(self, interaction: discord.Interaction):
        """Display bot statistics - mirrors JS utility.js stats."""
        import psutil
        import time
        
        # Bot stats
        uptime_seconds = int(time.time() - psutil.Process().create_time())
        uptime = format_duration(uptime_seconds * 1000)
        
        # Memory
        process = psutil.Process()
        mem_usage = process.memory_info().rss / 1024 / 1024
        
        guilds = len(self.bot.guilds)
        active_players = sum(1 for g in self.bot.guilds if g.voice_client)
        
        embed = discord.Embed(
            title="📊 Bot Statistics",
            color=discord.Color.from_str('#5865F2')
        )
        
        embed.add_field(name="⏱️ Uptime", value=uptime, inline=True)
        embed.add_field(name="💾 Memory", value=f"{mem_usage:.2f} MB", inline=True)
        embed.add_field(name="🏠 Servers", value=str(guilds), inline=True)
        embed.add_field(name="🎵 Active Players", value=str(active_players), inline=True)
        embed.add_field(name="📡 Python", value=platform.python_version(), inline=True)
        embed.add_field(name="💻 Platform", value=f"{platform.system()} {platform.machine()}", inline=True)
        
        # Lavalink stats if available
        nodes = wavelink.Pool.nodes
        if nodes:
            node = list(nodes.values())[0]
            embed.add_field(name="\u200B", value="**Lavalink Node**", inline=False)
            embed.add_field(name="Node", value=node.identifier, inline=True)
            embed.add_field(name="Status", value="🟢 Connected" if node.status.is_connected else "🔴 Disconnected", inline=True)
            
            if hasattr(node, 'stats') and node.stats:
                stats = node.stats
                embed.add_field(name="Players", value=str(stats.players), inline=True)
                embed.add_field(name="Playing", value=str(stats.playing_players), inline=True)
        
        await interaction.response.send_message(embed=embed)
    
    @app_commands.command(name="disconnect", description="Disconnect from voice channel")
    async def disconnect(self, interaction: discord.Interaction):
        """Disconnect the bot from voice."""
        player = cast(wavelink.Player, interaction.guild.voice_client)
        
        if not player:
            return await interaction.response.send_message("Not connected to voice.", ephemeral=True)
        
        await player.disconnect()
        await interaction.response.send_message("👋 Disconnected.")


async def setup(bot: commands.Bot):
    await bot.add_cog(Utility(bot))

File: src/commands/utility.js
import { Command, Declare, Embed, Options, createStringOption } from 'seyfert';
import os from 'os';
import fs from 'fs';
import path from 'path';
import { musicManager, ffmpegInfo, hardwareOptimization, execFileAsync } from '../services/MusicManager.js';
import { getVoiceConnection } from '@discordjs/voice';
import { isOwner } from '../utils/permissions.js';

function formatDuration(seconds) {
    if (!seconds || isNaN(seconds)) return '0:00';
    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    const secs = Math.floor(seconds % 60);
    if (hours > 0) {
        return `${hours}:${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}`;
    }
    return `${minutes}:${secs.toString().padStart(2, '0')}`;
}
@Declare({
    name: 'ping',
    description: 'Check bot latency'
})
export class PingCommand extends Command {
    async run(ctx) {
        const start = Date.now();
        try {
            await ctx.deferReply();
        } catch (e) {
            console.warn(`[PingCommand] Failed to defer interaction (likely timeout or unknown interaction):`, e.message || e);
            return;
        }
        const wsPing = ctx.client.gateway.latency;
        const latency = Date.now() - start;

        let proxyStatus = "⚪ Disabled";
        if (process.env.YOUTUBE_PROXY) {
            const proxyStart = Date.now();
            try {
                // Test connection to YouTube via the custom proxy (HEAD request, 3s timeout)
                await execFileAsync('curl', ['-s', '-I', '-x', process.env.YOUTUBE_PROXY, 'https://www.youtube.com', '--connect-timeout', '3'], { timeout: 4000 });
                const proxyLatency = Date.now() - proxyStart;
                proxyStatus = `🟢 ${proxyLatency}ms`;
            } catch(e) {
                proxyStatus = "🔴 Connection Failed (Offline/Invalid)";
            }
        }

        const embed = new Embed()
            .setTitle('🏓 Pong!')
            .setColor('#00FF00')
            .addFields(
                { name: 'Bot Latency', value: `${latency}ms`, inline: true },
                { name: 'Websocket Ping', value: `${wsPing}ms`, inline: true },
                { name: 'API Time', value: `${Date.now() - start}ms`, inline: true },
                { name: 'YouTube Proxy', value: proxyStatus, inline: false }
            );

        await ctx.editOrReply({ embeds: [embed] });
    }
}

// StatsCommand merged into DebugCommand

@Declare({
    name: 'debug',
    description: 'Show advanced debugging information for the bot and music player'
})
export class DebugCommand extends Command {
    async run(ctx) {
        try {
            await ctx.deferReply();
        } catch (e) {
            console.warn(`[DebugCommand] Failed to defer:`, e.message || e);
            return;
        }

        const queue = musicManager.getQueue(ctx.guildId);
        const connection = getVoiceConnection(ctx.guildId);

        const uptime = formatDuration(Math.floor(process.uptime()));
        const memUsage = (process.memoryUsage().heapUsed / 1024 / 1024).toFixed(2);
        const runtime = typeof Bun !== 'undefined' ? `Bun v${Bun.version}` : `Node ${process.version}`;

        let debugInfo = `**Bot Status:**\n`;
        debugInfo += `> **Shard ID:** \`${ctx.shardId}\`\n`;
        debugInfo += `> **Uptime:** \`${uptime}\`\n`;
        debugInfo += `> **Memory:** \`${memUsage} MB\`\n`;
        debugInfo += `> **Runtime:** \`${runtime}\` (${os.platform()} ${os.arch()})\n`;
        debugInfo += `> **Gateway Ping:** \`${ctx.client.gateway.latency}ms\`\n\n`;

        let ytdlpVersion = 'Error/Not Found';
        const ytdlpPath = process.env.YTDLP_PATH || 'yt-dlp';
        try {
            const { stdout } = await execFileAsync(ytdlpPath, ['--version']);
            ytdlpVersion = stdout.trim();
        } catch(e) {}

        debugInfo += `**Music Engine:**\n`;
        debugInfo += `> **FFmpeg:** \`v${ffmpegInfo.version}\` (${ffmpegInfo.type})\n`;
        debugInfo += `> **Path:** \`${ffmpegInfo.path}\`\n`;
        debugInfo += `> **yt-dlp:** \`v${ytdlpVersion}\`\n`;
        debugInfo += `> **Path:** \`${ytdlpPath}\`\n\n`;

        const hasCookies = fs.existsSync(path.join(process.cwd(), 'cookies.txt')) || fs.existsSync(path.join(process.cwd(), 'youtube-cookies.txt')) || !!process.env.YOUTUBE_COOKIES;
        let pluginStatus = 'Not Installed';
        const pipCommands = [
            process.env.YTDLP_PATH ? [process.env.YTDLP_PATH.replace('yt-dlp', 'pip'), ['show', 'yt-dlp-getpot-wpc']] : null,
            ['pip', ['show', 'yt-dlp-getpot-wpc']],
            ['pip3', ['show', 'yt-dlp-getpot-wpc']],
            ['python3', ['-m', 'pip', 'show', 'yt-dlp-getpot-wpc']],
            ['python', ['-m', 'pip', 'show', 'yt-dlp-getpot-wpc']]
        ].filter(Boolean);

        for (const [cmd, args] of pipCommands) {
            try {
                const { stdout } = await execFileAsync(cmd, args);
                if (stdout.includes('Version:')) {
                    pluginStatus = 'Active';
                    break;
                }
            } catch (e) {}
        }

        debugInfo += `**YouTube Bypasses:**\n`;
        debugInfo += `> **Cookies:** \`${hasCookies ? 'Loaded' : 'None'}\`\n`;
        debugInfo += `> **User-Agent:** \`${process.env.YOUTUBE_USER_AGENT ? 'Custom' : 'Default'}\`\n`;
        debugInfo += `> **PoToken Plugin:** \`${pluginStatus}\`\n`;

        const proxyUrl = process.env.YOUTUBE_PROXY;
        if (proxyUrl) {
            const isSocks = proxyUrl.toLowerCase().startsWith('socks');
            const proxyType = isSocks ? 'SOCKS5' : 'HTTP';
            const streamMode = isSocks ? 'Download-to-file (FFmpeg cannot use SOCKS)' : 'Direct stream via FFmpeg';
            debugInfo += `> **Proxy:** \`${proxyType}\` → \`${streamMode}\`\n`;
        } else {
            debugInfo += `> **Proxy:** \`None\`\n`;
        }
        debugInfo += `\n`;

        const dbEngine = (process.env.CLOUDFLARE_D1_TOKEN && process.env.CLOUDFLARE_D1_DATABASE_ID) 
            ? 'Cloudflare D1 (Serverless SQLite)' 
            : 'Local bun:sqlite (WAL Mode)';

        debugInfo += `**Storage & Hardware:**\n`;
        debugInfo += `> **Database Engine:** \`${dbEngine}\`\n`;
        debugInfo += `> **Threads:** \`${hardwareOptimization.threads}\`\n`;
        debugInfo += `> **Buffer:** \`${hardwareOptimization.bufferSizeMB} MB\`\n`;
        debugInfo += `> **Disk Type:** \`${hardwareOptimization.diskType.toUpperCase()}\`\n\n`;

        if (queue && connection) {
            debugInfo += `**Voice Connection:** ${connection.state.status}\n`;
            debugInfo += `**Audio Player:** ${queue.player.state.status}\n`;
            debugInfo += `**Queue Length:** ${queue.songs.length}\n`;
            debugInfo += `**Now Playing:** ${queue.playing ? 'Yes' : 'No'}\n`;
            debugInfo += `**Paused:** ${queue.paused ? 'Yes' : 'No'}\n`;
            debugInfo += `**Process Stream:** ${queue.ytdlpProcess ? 'Active' : 'None'}\n`;
            debugInfo += `**Resource:** ${queue.resource ? 'Active' : 'None'}\n`;

            const currentSong = queue.songs[0];
            if (currentSong && currentSong._ytDlpData) {
                const streamData = currentSong._ytDlpData;
                const bitrateVal = streamData.abr || streamData.tbr;
                const bitrateStr = bitrateVal ? `${typeof bitrateVal === 'number' ? bitrateVal.toFixed(1) : bitrateVal} kbps` : 'N/A';
                
                // Source Platform & Protocol
                const platformName = currentSong.sourceType ? currentSong.sourceType.toUpperCase() : (streamData.extractor ? streamData.extractor.toUpperCase() : 'UNKNOWN');
                const targetProtocol = (currentSong.sourceType === 'radio' || currentSong.isRadio) ? 'HTTP Stream' : 'HTTPS/QUIC';
                
                // Proxy Route
                const proxyScheme = proxyUrl ? (proxyUrl.includes('://') ? proxyUrl.split('://')[0].toUpperCase() : 'PROXY') : '';
                const proxyLabel = proxyUrl ? `${proxyScheme} Proxy` : 'Direct (No Proxy)';

                debugInfo += `\n**Stream Info:**\n`;
                debugInfo += `> **Engine Mode:** \`${streamData._processingMode || 'N/A'}\`\n`;
                debugInfo += `> **Transport:** \`${streamData._transportMode || 'N/A'}\`\n`;
                debugInfo += `> **Source Platform:** \`${platformName} (${targetProtocol})\`\n`;
                debugInfo += `> **Proxy Route:** \`${proxyLabel}\`\n`;
                debugInfo += `> **Audio Delivery:** \`UDP (Discord Voice Socket)\`\n`;
                debugInfo += `> **Source Codec:** \`${streamData.acodec || 'N/A'}\`\n`;
                debugInfo += `> **Bitrate:** \`${bitrateStr}\`\n`;
                debugInfo += `> **Sample Rate:** \`${streamData.asr ? `${streamData.asr} Hz` : 'N/A'}\`\n`;
                const codecExt = (streamData.acodec && streamData.ext) ? `${streamData.acodec}/${streamData.ext}` : (streamData.ext || 'N/A');
                debugInfo += `> **Format ID:** \`${streamData.format_id || 'N/A'} (${codecExt})\`\n`;
            }
        } else {
            debugInfo += `**Voice Connection:** None\n`;
            debugInfo += `**Queue:** Not initialized\n`;
        }
        const embed = new Embed()
            .setTitle('🛠️ Debug Info')
            .setDescription(debugInfo)
            .setColor('#ED4245');

        await ctx.editOrReply({ embeds: [embed] });
    }
}

const helpOptions = {
    category: createStringOption({
        description: 'Show help for a specific category',
        choices: [
            { name: '🎶 Playing Music', value: 'playing' },
            { name: '⏯️ Playback', value: 'playback' },
            { name: '📋 Queue', value: 'queue' },
            { name: '🔧 Utility', value: 'utility' }
        ]
    })
};

@Declare({
    name: 'help',
    description: 'Show help for available commands.'
})
@Options(helpOptions)
export class HelpCommand extends Command {
    async run(ctx) {
        const category = ctx.options.category;

        const categoryData = {
            playing: {
                title: '🎶 Playing Music Commands',
                fields: [
                    { name: '`/play <query>`', value: 'Play a track or playlist from YouTube, Spotify, SoundCloud, Bandcamp, Vimeo, Twitch, or direct audio link.' },
                    { name: '`/forceplay <query>`', value: 'Force play a song immediately, clearing or interrupting current playback.' },
                    { name: '`/playnext <query>`', value: 'Add a track to the very front of the queue to play next.' },
                    { name: '`/search <query> [platform]`', value: 'Search interactively on YouTube or SoundCloud and pick a result.' },
                    { name: '`/radio <preset|url>`', value: 'Stream live internet radio stations or built-in presets (lofi, jazz, synthwave, etc.).' }
                ]
            },
            playback: {
                title: '⏯️ Playback Control Commands',
                fields: [
                    { name: '`/pause`', value: 'Pause current audio playback.' },
                    { name: '`/resume`', value: 'Resume paused audio playback.' },
                    { name: '`/stop`', value: 'Stop playback, clear queue, and leave voice channel.' },
                    { name: '`/volume <0-200>`', value: 'Adjust playback volume percentage.' },
                    { name: '`/seek <timestamp>`', value: 'Seek to a specific timestamp in the current song (e.g. `1:30`, `90`).' },
                    { name: '`/replay`', value: 'Restart the currently playing song from the beginning.' },
                    { name: '`/previous`', value: 'Play the previously played song from history.' }
                ]
            },
            queue: {
                title: '📋 Queue Commands',
                fields: [
                    { name: '`/queue [page]`', value: 'View the paginated song queue.' },
                    { name: '`/nowplaying` / `/np`', value: 'Show interactive Now Playing embed with live progress bar.' },
                    { name: '`/shuffle`', value: 'Randomize the order of songs in the queue.' },
                    { name: '`/clear`', value: 'Remove all upcoming songs from the queue.' },
                    { name: '`/remove <position>`', value: 'Remove a specific song by its queue position.' },
                    { name: '`/move <from> <to>`', value: 'Reorder a song from one position to another.' },
                    { name: '`/jump <position>` / `/skipto`', value: 'Jump directly to a song in the queue.' },
                    { name: '`/skip` / `/voteskip`', value: 'Skip the current track (or start a vote skip).' }
                ]
            },
            utility: {
                title: '🔧 Utility Commands',
                fields: [
                    { name: '`/help [category]`', value: 'Display command guide and category breakdowns.' },
                    { name: '`/ping`', value: 'Check bot latency, gateway ping, and YouTube proxy status.' },
                    { name: '`/debug`', value: 'View detailed server hardware, yt-dlp, FFmpeg, and voice stream diagnostics.' },
                    { name: '`/lyrics [query]`', value: 'Search and display song lyrics.' }
                ]
            }
        };

        if (category && categoryData[category]) {
            const data = categoryData[category];
            const embed = new Embed()
                .setColor('#5865F2')
                .setTitle(data.title)
                .addFields(data.fields)
                .setFooter({ text: 'Powered by Seyfert' });
            await ctx.write({ embeds: [embed] });
            return;
        }

        const helpEmbed = new Embed()
            .setColor('#5865F2')
            .setTitle('🎵 Music Bot Commands')
            .addFields(
                { name: '🎶 Playing Music', value: '`/play` `/forceplay` `/playnext` `/search` `/radio`', inline: false },
                { name: '⏯️ Playback Control', value: '`/pause` `/resume` `/stop` `/volume` `/seek` `/replay` `/previous`', inline: false },
                { name: '📋 Queue Management', value: '`/queue` `/nowplaying` `/shuffle` `/clear` `/remove` `/move` `/jump` `/skip` `/voteskip`', inline: false },
                { name: '🔧 Utility', value: '`/help` `/ping` `/debug` `/lyrics`', inline: false }
            )
            .setFooter({ text: 'Use /help category:<name> for detailed usage of any section' });

        await ctx.write({ embeds: [helpEmbed] });
    }
}
